containers/agent/checksum_discovery.go

```diff
@@ -0,0 +1,171 @@
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	metal3v1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+	"github.com/spf13/cobra"
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+var (
+	checksumStartCmd = &cobra.Command{
+		Use:   "checksum-discovery",
+		Short: "Start Checksum Discovery Agent",
+		Long:  "",
+		Run:   runChecksumStartCmd,
+	}
+
+	checksumStartOpts struct {
+		kubeconfig          string
+		osImageDir          string
+		provServerName      string
+		provServerNamespace string
+	}
+)
+
+func init() {
+	rootCmd.AddCommand(checksumStartCmd)
+	checksumStartCmd.PersistentFlags().StringVar(&checksumStartOpts.osImageDir, "os-image-dir", "", "OS image directory on the associated host")
+	checksumStartCmd.PersistentFlags().StringVar(&checksumStartOpts.provServerName, "prov-server-name", "", "Provisioning server resource name")
+	checksumStartCmd.PersistentFlags().StringVar(&checksumStartOpts.provServerNamespace, "prov-server-namespace", "", "Provisioning server resource namespace")
+}
+
+func runChecksumStartCmd(_ *cobra.Command, _ []string) {
+	var err error
+	err = flag.Set("logtostderr", "true")
+	if err != nil {
+		panic(err.Error())
+	}
+
+	flag.Parse()
+
+	glog.V(0).Info("Starting ChecksumDiscoveryAgent")
+
+	if checksumStartOpts.osImageDir == "" {
+		dir, ok := os.LookupEnv("OS_IMAGE_DIR")
+		if !ok || dir == "" {
+			glog.Fatalf("os-image-dir is required")
+		}
+		checksumStartOpts.osImageDir = dir
+	}
+
+	if checksumStartOpts.provServerName == "" {
+		name, ok := os.LookupEnv("PROV_SERVER_NAME")
+		if !ok || name == "" {
+			glog.Fatalf("prov-server-name is required")
+		}
+		checksumStartOpts.provServerName = name
+	}
+
+	if checksumStartOpts.provServerNamespace == "" {
+		name, ok := os.LookupEnv("PROV_SERVER_NAMESPACE")
+		if !ok || name == "" {
+			glog.Fatalf("prov-server-namespace is required")
+		}
+		checksumStartOpts.provServerNamespace = name
+	}
+
+	var config *rest.Config
+	kubeconfig := os.Getenv("KUBECONFIG")
+	if kubeconfig != "" {
+		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+	} else {
+		// creates the in-cluster config
+		config, err = rest.InClusterConfig()
+	}
+
+	if err != nil {
+		panic(err.Error())
+	}
+
+	dClient := dynamic.NewForConfigOrDie(config)
+
+	provServerClient := dClient.Resource(openstackProvisionServerGVR)
+
+	checksumFileName := ""
+	var checksumType metal3v1.ChecksumType
+
+	// First get the checksum data
+	dir, err := os.Open(checksumStartOpts.osImageDir)
+
+	if err != nil {
+		panic(err.Error())
+	}
+
+	items, err := dir.Readdirnames(0)
+	_ = dir.Close()
+
+	if err != nil {
+		panic(err.Error())
+	}
+
+	for _, item := range items {
+		// Crude mechanism for detecting both the checksum file and its type
+		if strings.Contains(item, "md5") {
+			checksumFileName = item
+			checksumType = metal3v1.MD5
```

Comment:
Should we just look for sha256 because that's the only one we validate when copying from the image? https://github.com/openstack-k8s-operators/edpm-image-builder/blob/main/copy_out.sh#L62C1-L62C13

Also, dowstream they're hard-coding the checksum filename https://pkgs.devel.redhat.com/git/containers/edpm-hardened-uefi/tree/Dockerfile?h=rhoso-18.0-rhel-9#n8.  So maybe having the `osImageCheksum`  field that defaults to `edpm-hardened-uefi.qcow2.sha256` would make it simpler(?).

Reply:
> which doesn't work when the checksum is generated as .sha256sum

We can merge this but I think we're probably redundantly catering of someone generating checksum file with whatever name. The same way we could be expected to use the whatever qcow2 image in `osImageDir`.
We won't be supporting anything other than sha256 when we generate/validate the checksum as in edpm-image-builder. I was thinking of something simpler like https://github.com/openstack-k8s-operators/openstack-baremetal-operator/compare/main...rabi:openstack-baremetal-operator:checksum which aligns with a default vs custom (user provided) qcow image name.

Reply:
As I did not hear from you on my proposal, going ahead and merging this. We can rejig it in a later date.

Reply:
Sorry, I somehow missed this one during the GA release frenzy.  Just took a look at your proposal and it definitely looks cleaner.  If we're sure we'll only ever support `sha256`, then I'm on board with refactoring it as you've demonstrated.

Reply:
My point was we will always build the os container image which would be part of the release payload and if it has to change in the future other things can. The current assumption that the checksum file name has the checksumType in it won't probably be easy to ensure, if customers are allowed to build their os container image with checksum file in it ;)